horovod/common/operations.cc

```diff
@@ -1109,6 +1132,10 @@ void BackgroundThreadLoop(HorovodGlobalState& state, MPIContext& ctx) {
     state.message_table = std::unique_ptr<MessageTable>(new MessageTable());
   }
 
+#if COMPILE_WITH_GLOO
+  gloo_context.InitializeFromMPI(ctx.mpi_comm);
```

Comment:
We should also check if Gloo is enabled here to avoid creating Gloo communicator we aren't going to use.

One thing we can do is add a field to the GlooContext called `enabled_cpu_operations`.  We can set this once during initialization at the same place in `BackgroundThreadLoop` we check the other environment variables.

Then for initialization and `Finalize()` below, we just need to check against `gloo_context.enabled_cpu_operations`.

Reply:
I have created two flags indicating whether gloo will be used for the data layer and the control layer. And the initialization and finalization functions will check these flags to determine whether we should actually initialize or finalize.